### PR TITLE
Build benchmarks with upstream pytorch

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
WIP

Reference: #1911

### CODE REVIEW

**Comment by ZzEeKkAa** [comment]
@Wheest @ESI-SYD @chengjunlu could you please take a look at this DRAFT changes. I've converted it to work with upstream pytorch and made it work, so it compiles and produce number. But I don't have expertise in the product and how legit those changes are and if they negatively affects results. Could you review conceptually if this may take place. If so, I'll prepare it for the code review and make it production ready.

File: benchmarks/triton_kernels_benchmark/benchmark_testing.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def do_bench(fn, warmup=25, rep=100, grad_to_none=None, quantiles=None, fast_flu
     for _ in range(n_warmup):
         fn()
     # Benchmark
-    with torch.autograd.profiler_legacy.profile(True, use_xpu=True) as prof:
+    with torch.profiler.profile() as prof:
```
Comment (chengjunlu) [inline benchmarks/triton_kernels_benchmark/benchmark_testing.py:71]: Need to refer the Torch profiling step to capture the XPU kernel performance.

Comment (ZzEeKkAa) [reply benchmarks/triton_kernels_benchmark/benchmark_testing.py:71]: From the function description it should capture it whenever it is possible. I've set config explicitly hoping it will fail if XPU tracing is unavailable.

File: benchmarks/triton_kernels_benchmark/benchmark_testing.py [hunk 2]
```diff
@@ -86,7 +86,7 @@ def do_bench(fn, warmup=25, rep=100, grad_to_none=None, quantiles=None, fast_flu
         # Record clocks
         synchronize()
 
-    profiling_func_filter = filter(lambda x: x.name.startswith("__profile_kernel_of_func"), prof.function_events)
+    profiling_func_filter = filter(lambda x: x.name.startswith("__profile_kernel_of_func"), prof.events())
```
[the remark this answers is not in the document]
Comment (ZzEeKkAa) [reply benchmarks/triton_kernels_benchmark/benchmark_testing.py:89]: Could you provide some links what are you referring?

**Comment by ZzEeKkAa** [comment]
> Basically looks good to me.
> 
> Need to confirm that the Torch profiler (Kineto) can capture the Triton kernel performance. Here is the steps to build the Torch with Kineto XPU supporting.
> 
> 1. Install the Intel PTI package: l_intel-pti-dev_p_0.9.0.38
> 2. Source the PTI env: source ${HOME}/intel/oneapi/pti/latest/env/vars.sh
> 3. Build the stock Pytorch.

I've build pytorch with PTI and was managed to get numbers out of gemm benchmark, but I don't have enough domain knowledge to confirm that they makes sense.

File: benchmarks/triton_kernels_benchmark/benchmark_testing.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chengjunlu) [reply benchmarks/triton_kernels_benchmark/benchmark_testing.py:95]: I think you have already got the performance from the PTI. It is done.

File: benchmarks/triton_kernels_benchmark/benchmark_testing.py (hunk 1 above)
Comment (chengjunlu) [reply benchmarks/triton_kernels_benchmark/benchmark_testing.py]: I thought we need to explicitly specified the XPU just as:
```
with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.XPU]) as prof:
```

I think it is ok if the XPU kernel performance is enabled by default.

**Comment by chengjunlu** [comment]
> I've build pytorch with PTI and was managed to get numbers out of gemm benchmark, but I don't have enough domain knowledge to confirm that they makes sense.

We can compare the benchmark reports to the nightly performance reports.

File: benchmarks/triton_kernels_benchmark/benchmark_driver.py [hunk 3]
```diff
@@ -10,15 +10,15 @@
 from triton.runtime.build import _build, quiet
 
 import torch
-import intel_extension_for_pytorch
+# import intel_extension_for_pytorch
```
Comment (chengjunlu) [inline benchmarks/triton_kernels_benchmark/benchmark_driver.py:13]: We can simply remove this file as long as Kineto works by removing the usage of this file at here:
https://github.com/intel/intel-xpu-backend-for-triton/blob/c7c60809cccfee24a650708712d9cc5aecf3db9f/benchmarks/triton_kernels_benchmark/__init__.py#L12

Comment (ZzEeKkAa) [reply benchmarks/triton_kernels_benchmark/benchmark_driver.py:13]: `benchmark_driver` is the file that you are saying to remove. If I remove both file and the highlighted line, I'm getting no numbers.

File: benchmarks/xetla_kernel/python_main.cpp [hunk 4]
```diff
@@ -99,7 +107,7 @@ at::Tensor bf16_stream_k_gemm(const at::Tensor &a, const at::Tensor &b,
   auto queue = get_current_sycl_queue();
   auto evt = stream_k_gemm_run(a.data_ptr(), b.data_ptr(), c.data_ptr(),
                                acc.data_ptr(), cnt.data_ptr(), queue);
-  xpu::profiler_record("xetla kernel", evt);
+  //   xpu::profiler_record("xetla kernel", evt);
```
Comment (ZzEeKkAa) [inline benchmarks/xetla_kernel/python_main.cpp:110]: I've also remove calls to `xpu::profiler_record("xetla kernel", evt);`. What it was doing? Why is it working without this usage?

Comment (ZzEeKkAa) [reply benchmarks/xetla_kernel/python_main.cpp:110]: @chengjunlu

Comment (chengjunlu) [reply benchmarks/xetla_kernel/python_main.cpp:78]: There are two profilers in the Torch. `torch.autograd.profiler_legacy.profile` and `torch.profiler.profile`.

The `xpu::profiler_record` is to add the SYCL event to the first profile and the legacy profile can get the Triton kernel performance statistics for using.

The second profiler is based on Kineto and PTI. If the Kineto can get the Triton kernel performance thru the PTI, then the function `xpu::profiler_record` is no longer required.

**Comment by chengjunlu** [comment]
@ZzEeKkAa , The Kineto has issue in tracing the Triton kernel performance. The `triton.testing.do_bench` is used in the benchmarks for Pytorch 2.5.

We can use the benchmarks `do_bench` after the Kineto issue has been fixed.

File: benchmarks/triton_kernels_benchmark/benchmark_driver.py (hunk 3 above)
Comment (chengjunlu) [reply benchmarks/triton_kernels_benchmark/benchmark_driver.py]: I met the same issue as yours. If we remove the `RECORD_FUNCTION` in the launcher, the Kineto cannot capture any Triton kernel performance.

I am not sure whether the time captured with the benchmark's launcher is correct. 

We use the Triton `do_bench` for now. Let's revisit the Kineto when the Kineto issue was solved.

File: benchmarks/xetla_kernel/CMakeLists.txt [hunk 5]
```diff
@@ -29,13 +29,11 @@ endif()
 add_library(xetla_kernel SHARED python_main.cpp)
 set_target_properties(xetla_kernel PROPERTIES PREFIX "")
 target_compile_options(xetla_kernel PRIVATE "-fPIC")
-target_compile_options(xetla_kernel PRIVATE "-fsycl")
+target_compile_options(xetla_kernel PRIVATE "-fsycl" "-fpreview-breaking-changes")
```
Comment (chengjunlu) [inline benchmarks/xetla_kernel/CMakeLists.txt:32]: @ZzEeKkAa The changes is required to run the XeTLA kernel in the benchmarks.

But do you have any idea why this flag is required?
https://github.com/intel/llvm/blob/283073a4f21aa44203fd49e25b4ff0379c91b29a/sycl/doc/UsersManual.md?plain=1#L429

**Comment by ESI-SYD** [comment]
Can we update to setup upstream pytorch? Then we can run it in Actions.
https://github.com/intel/intel-xpu-backend-for-triton/blob/386ae76aebefa156d858f8cf6f58247764067340/.github/workflows/triton-benchmarks.yml#L52-L56

**Review by chengjunlu — APPROVED** [review]

**Review by chengjunlu — APPROVED** [review]

File: .github/workflows/triton-benchmarks.yml [hunk 6]
```diff
@@ -49,12 +57,28 @@ jobs:
...
         uses: ./.github/actions/setup-pytorch
+        with:
+          repository: pytorch/pytorch
+          ref: ${{ inputs.pytorch_ref }}
 
       - name: Setup IPEX
+        if: ${{ inputs.install_ipex }}
         uses: ./.github/actions/setup-ipex
 
+      - name: Setup no-op IPEX
```
Comment (whitneywhtsang) [inline .github/workflows/triton-benchmarks.yml:78]: Given that we are removing `import intel_extension_for_pytorch` in this PR, do we still need to setup no-op IPEX?

File: benchmarks/key_benchmarks/triton_kernels_benchmark/benchmark_testing.py
Comment (whitneywhtsang) [inline benchmarks/key_benchmarks/triton_kernels_benchmark/benchmark_testing.py:1]: What's the performance impact, now that we change back to use elapsed time?

**Comment by ESI-SYD** [comment]
[Upstream PT(9f1945a3bc1defa56c01cb1fbdfdd3a938e8c025)](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/10609254853/job/29404643614)
vs [pt+ipex](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/10605751168/job/29395177166)
: Has performance lost

[preprocessed-data-upstream-ipex.csv](https://github.com/user-attachments/files/16809422/preprocessed-data-upstream-ipex.csv)
[performance-plot-upstream-xetla-tflops-upstream-ipex-xetla-tflops-ipex.pdf](https://github.com/user-attachments/files/16809400/performance-plot-upstream-xetla-tflops-upstream-ipex-xetla-tflops-ipex.pdf)
[performance-plot-upstream-triton-tflops-upstream-ipex-triton-tflops-ipex.pdf](https://github.com/user-attachments/files/16809401/performance-plot-upstream-triton-tflops-upstream-ipex-triton-tflops-ipex.pdf)

File: .github/workflows/triton-benchmarks.yml (hunk 6 above)
Comment (ESI-SYD) [reply .github/workflows/triton-benchmarks.yml:78]: Yes, can removed

**Comment by whitneywhtsang** [comment]
> [Upstream PT(9f1945a3bc1defa56c01cb1fbdfdd3a938e8c025)](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/10609254853/job/29404643614) vs [pt+ipex](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/10605751168/job/29395177166) : Has performance lost
> 
> [preprocessed-data-upstream-ipex.csv](https://github.com/user-attachments/files/16809422/preprocessed-data-upstream-ipex.csv) [performance-plot-upstream-xetla-tflops-upstream-ipex-xetla-tflops-ipex.pdf](https://github.com/user-attachments/files/16809400/performance-plot-upstream-xetla-tflops-upstream-ipex-xetla-tflops-ipex.pdf) [performance-plot-upstream-triton-tflops-upstream-ipex-triton-tflops-ipex.pdf](https://github.com/user-attachments/files/16809401/performance-plot-upstream-triton-tflops-upstream-ipex-triton-tflops-ipex.pdf)

Not sure how to read the result, looks like it failed to run? 
If there is performance lost, do you know why? Is the relative performance still the same between XeTLA and Triton?

**Comment by ESI-SYD** [comment]
I updated this PR, skipped upstream torch sdpa failure to get successful [run](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/10629513475/job/29466503753) to unblock Fa benchmark.

Take this one as [reference](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/10623940969/job/29451334896) (with IPEX)

From reports generated by [script](https://github.com/intel/intel-xpu-backend-for-triton/pull/2060) and focus on cases whose relative difference value far from zero.

```bash
rm -rf performance* preprocess* ipex/ upstream/ && python compare_runs.py -d "ipex:10623940969" -n "upstream:10629513475" --triton-benchmark
```

**Difference summary:**
1. Ratio of `Triton`/ `XeTLA` keep same except for attention.
image
2. Attention `XeTLA` cases degraded (reproduced, needs fix)
image
3. Both `Triton` and `XeTLA` softmax cases degraded, so ratio not changed.

File: benchmarks/key_benchmarks/triton_kernels_benchmark/benchmark_testing.py
Comment (whitneywhtsang) [reply benchmarks/key_benchmarks/triton_kernels_benchmark/benchmark_testing.py:1]: Answered in https://github.com/intel/intel-xpu-backend-for-triton/pull/1905#issuecomment-2320701513.

**Review by etiotto — CHANGES_REQUESTED** [review]
I am under the impression the benchmark FA code cannot be run using PyTorch and can only be run correctly using IPEX. This suggests that we *cannot* remove the dependency on IPEX for the benchmark codes until pytorch fixes the functional problem affecting FA.  For this reason I am blocking the PR from merging.  We cannot remove verification code from the FA benchmark because doing so would prevent proper functional checking for PRs that might modify the FA codegen.

**Comment by kwasd** [comment]
(required for #1933)

**Comment by whitneywhtsang** [comment]
Blocked by https://github.com/intel/intel-xpu-backend-for-triton/issues/2042, converting to draft.

**Comment by anmyachev** [comment]
> I updated this PR, skipped upstream torch sdpa failure to get successful [run](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/10629513475/job/29466503753) to unblock Fa benchmark.
> 
> Take this one as [reference](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/10623940969/job/29451334896) (with IPEX)
> 
> From reports generated by [script](https://github.com/intel/intel-xpu-backend-for-triton/pull/2060) and focus on cases whose relative difference value far from zero.
> 
> ```shell
> rm -rf performance* preprocess* ipex/ upstream/ && python compare_runs.py -d "ipex:10623940969" -n "upstream:10629513475" --triton-benchmark
> ```
> 
> **Difference summary:**
> 
> 1. Ratio of `Triton`/ `XeTLA` keep same except for attention.
> 2. Attention `XeTLA` cases degraded (reproduced, needs fix)
> 3. Both `Triton` and `XeTLA` softmax cases degraded, so ratio not changed.

I still see `RECORD_FUNCTION("xetla softmax", {input});` (and similar) function calls. Can this affect the performance?
